### DISCUSSION TITLE
Database version aware `with_variant`

### DISCUSSION THREAD

**Discussion by cofin** [sqlalchemy/sqlalchemy#10534] [Usage Questions] [answered]
Hi - 

Is there a way to utilize the `with_variant` option and include the database version number as part of the criteria?

I would like to make a custom JSON type for the new Oracle 23c JSON type.  However, I already have a JSON type for all versions >12.

You can see the current [code here](https://github.com/jolt-org/advanced-alchemy/blob/main/advanced_alchemy/types.py#L173)

It would be great to be able to specify something along the lines of this logic:
```python
JsonB = (
    _JSON().with_variant(PG_JSONB, "postgresql").with_variant(ORA_JSONB, "oracle", version<=23).with_variant(ORA_NATIVE_JSONB, "oracle", version>=23).with_variant(PG_JSONB, "cockroachdb")
)
```

I suspect that it's not currently possible, and I'd need to move this logic to the `ORA_JSONB` custom type, but I wanted to confirm this was the best approach before attempting.

**Comment by zzzeek** [comment, marked as answer]
it's not and I think you might be better off building a TypeDecorator for at least the oracle part here which gives you more control over this kind of thing in the [load_dialect_impl](https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator.load_dialect_impl) method, where the dialect has "server_version_info".  however, you wont have server_version_info until the database has connected at least once.  it should be fine for "impl" as "impl" is not accessed until a SQL statement is compiled, so as long as that's within an excecute call, it would work out.

**Reply by cofin** [reply]
Thanks for the quick reply.  I'll give this a try.
